### PR TITLE
Improve debugging experience by allowing "human readable" output

### PR DESCRIPTION
**Pull request by Geod24**
As I am working on re-adding the integration test once and for all (and making sure it tells us why it fails next time), I found it very hard to debug Agora. This adds a few QoL improvements which should make debugging much easier in the future. Some of them might prove controversial, so feel free to go ahead and shoot.

This also includes a few improvements for the CI. If this PR lingers for too long, or if it's too much to review, I can split them out.

### CODE REVIEW

**Comment by Geod24** [comment]
Side note: Currently the logging is very inefficient. It should take a pointer, but it looks like there's some memory corruption happening with `std.algorithm : map`. I choose to go the easy way and copy everything for the time being.

File: source/agora/utils/HumanReadable.d [hunk 1]
```diff
@@ -0,0 +1,284 @@
+/*******************************************************************************
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: `reduce` => `improves` in the commit message imo. Unless you also means it makes it more compact.

File: source/agora/utils/HumanReadable.d [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    Defines method to format types to their human readable representation
+
+    This approach allows to take the string formatting out of the type,
+    simplifying the implementation and reducing dependencies.
+    It is assumed that a human-readable representation will not rely on
+    non-visible data, or data that cannot be accessed without mutation.
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: Could you elaborate what "non-visible data" means?

File: source/agora/utils/HumanReadable.d [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+import agora.common.crypto.Key;
+import agora.consensus.data.Block;
+import agora.consensus.data.Transaction;
+
+import std.algorithm;
+import std.format;
+import std.range;
+
+/// Ditto
+public auto humanReadable (T) (const ref T input)
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: In many libraries this is called "pretty" printing. So something like `toPretty` is short and concise. But you can keep `humanReadable` too if you want. I'm just thinking of making this function shorter to write.

File: source/agora/utils/HumanReadable.d [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        this.value = r;
+    }
+
+    public void toString (scope void delegate(scope const(char)[]) @safe sink) @safe
+    {
+        // Only format `0xABCD..EFGH`
+        enum StartUntil = 6;
+        enum EndFrom    = Hash.StringBufferSize - 4;
+        size_t count;
+        scope void delegate(scope const(char)[]) @safe wrapper = (scope data) @safe {
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: ~Egyptian style braces?~

Comment (AndrejMitrovic) [reply source/agora/utils/HumanReadable.d]: Actually I see it in the style guide. Very well, carry on.

File: source/agora/utils/HumanReadable.d [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        this.value.toString(wrapper);
+    }
+}
+
+@safe unittest
+{
+    static immutable Hash SomeHash =
+        "0x000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f"
+        ~ "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f";
+    assert(format("%s", HashFmt(SomeHash)) == "0x0000...e26f");
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: Do you think we could make it all-caps? I'm thinking it might improve the readability.

Comment (AndrejMitrovic) [reply source/agora/utils/HumanReadable.d]: Ah, actually you are using all-caps for public keys. It makes sense to keep this lowercase.

File: source/agora/utils/HumanReadable.d [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    {
+        formattedWrite(sink, "%s(%s)",
+            AmountFmt(this.value.value), PubKeyFmt(this.value.address));
+    }
+}
+
+@safe unittest
+{
+    Output output;
+    assert(format("%s", OutputFmt(output)) == "0.0000,000(GAAA...AWHF)");
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: Hmm.. I would prefer Address(Value), ordering by importance from left to right. What do you think?

File: source/agora/utils/HumanReadable.d [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        formattedWrite(sink, "%s, Transactions: %d\n",
+            BlockHeaderFmt(this.value.header), this.value.txs.length);
+        formattedWrite(sink, "%(%s\n%)", this.value.txs.map!(v => TransactionFmt(v)));
+    }
+}
+
+@safe unittest
+{
+    import agora.consensus.Genesis;
+    static immutable GenesisStr = `Height: 0, Prev: 0x0000...0000, Root: 0x893a...d115, Transactions: 1
```
Comment (AndrejMitrovic) [inline source/agora/utils/HumanReadable.d]: Could you add a test with multiple transactions? I want to see how it formats.

File: source/agora/utils/HumanReadable.d (hunk 6 above)
Comment (Geod24) [reply source/agora/utils/HumanReadable.d]: SGTM

File: source/agora/consensus/Validation.d [hunk 8]
```diff
@@ -33,23 +33,24 @@ public alias UTXOFinder = scope const(Output)* delegate (Hash hash, size_t index
...
 
     Return:
-        Return true if this transaction is verified.
+        `null` if the transaction is valid, a string explaining the reason it
+        is invalid otherwise.
 
 *******************************************************************************/
 
-public bool isValid (const Transaction tx, UTXOFinder findUTXO)
+public string invalidReason (const Transaction tx, UTXOFinder findUTXO)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/Validation.d]: I dislike the name of this function. It seems to imply (on first glance) that the transaction/block is already invalid. And it's also not a verb, and hard to memorize. `validate` or `checkValid` or something like that might be better, but I can't suggest a really good name for this..

**Comment by AndrejMitrovic** [comment]
The rest LGTM.

File: source/agora/utils/HumanReadable.d (hunk 2 above)
Comment (Geod24) [reply source/agora/utils/HumanReadable.d]: `private` data. Because nowadays D is using "visibility" rather than "protection".

Comment (Geod24) [reply source/agora/utils/HumanReadable.d]: Well I'll add it to the docs

File: source/agora/utils/HumanReadable.d (hunk 1 above)
Comment (Geod24) [reply source/agora/utils/HumanReadable.d]: Yes, the point is not so much to have something "sensible", but something that doesn't take a whole screen.

**Comment by Geod24** [comment]
Fixed all comments

**Comment by AndrejMitrovic** [comment]
Looks good. I wonder if it would be useful to also output a Transaction's Hash in the output for debugging purposes?

**Comment by Geod24** [comment]
> Looks good. I wonder if it would be useful to also output a Transaction's Hash in the output for debugging purposes?

Probably very, same for blocks.

**Comment by Geod24** [comment]
Green

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**